# New Baseline Exploration

## Read in data

Our goal here is to read in the data for each file in the data directory.

In [1]:
# imports
from __future__ import print_function
from __future__ import division
import numpy as np
import os
from matplotlib import pyplot as plt
import pandas as pd
import seaborn as sns
import scipy.stats as stats

from sklearn import *
import sklearn
import pandas as pd
import numpy as np
import xgboost as xgb

C:\Users\jehun\Anaconda3\lib\site-packages\sklearn\cross_validation.py:44: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)
C:\Users\jehun\Anaconda3\lib\site-packages\sklearn\grid_search.py:43: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. This module will be removed in 0.20.
  DeprecationWarning)
C:\Users\jehun\Anaconda3\lib\site-packages\sklearn\lda.py:6: DeprecationWarning: lda.LDA has been moved to discriminant_analysis.LinearDiscriminantAnalysis in 0.17 and will be removed in 0.19
  "in 0.17 and will be removed in 0.19", DeprecationWarning)
C:\Users\jehun\Anacon

In [23]:
base_dir = 'D:\Backups\KaggleJul2017\TSA'
APS_FILE_NAME = 'D:\\Backups\\KaggleJul2017\\TSA\\stage1_aps.tar\\0a27d19c6ec397661b09f7d5998e0b14.aps'

In [17]:
#----------------------------------------------------------------------------------
# read_header(infile):  takes an aps file and creates a dict of the data
#
# infile:               an aps file
#
# returns:              all of the fields in the header
#----------------------------------------------------------------------------------

def read_header(infile):
    # declare dictionary
    h = dict()
    
    with open(infile, 'r+b') as fid:

        h['filename'] = b''.join(np.fromfile(fid, dtype = 'S1', count = 20))
        h['parent_filename'] = b''.join(np.fromfile(fid, dtype = 'S1', count = 20))
        h['comments1'] = b''.join(np.fromfile(fid, dtype = 'S1', count = 80))
        h['comments2'] = b''.join(np.fromfile(fid, dtype = 'S1', count = 80))
        h['energy_type'] = np.fromfile(fid, dtype = np.int16, count = 1)
        h['config_type'] = np.fromfile(fid, dtype = np.int16, count = 1)
        h['file_type'] = np.fromfile(fid, dtype = np.int16, count = 1)
        h['trans_type'] = np.fromfile(fid, dtype = np.int16, count = 1)
        h['scan_type'] = np.fromfile(fid, dtype = np.int16, count = 1)
        h['data_type'] = np.fromfile(fid, dtype = np.int16, count = 1)
        h['date_modified'] = b''.join(np.fromfile(fid, dtype = 'S1', count = 16))
        h['frequency'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['mat_velocity'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['num_pts'] = np.fromfile(fid, dtype = np.int32, count = 1)
        h['num_polarization_channels'] = np.fromfile(fid, dtype = np.int16, count = 1)
        h['spare00'] = np.fromfile(fid, dtype = np.int16, count = 1)
        h['adc_min_voltage'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['adc_max_voltage'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['band_width'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['spare01'] = np.fromfile(fid, dtype = np.int16, count = 5)
        h['polarization_type'] = np.fromfile(fid, dtype = np.int16, count = 4)
        h['record_header_size'] = np.fromfile(fid, dtype = np.int16, count = 1)
        h['word_type'] = np.fromfile(fid, dtype = np.int16, count = 1)
        h['word_precision'] = np.fromfile(fid, dtype = np.int16, count = 1)
        h['min_data_value'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['max_data_value'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['avg_data_value'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['data_scale_factor'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['data_units'] = np.fromfile(fid, dtype = np.int16, count = 1)
        h['surf_removal'] = np.fromfile(fid, dtype = np.uint16, count = 1)
        h['edge_weighting'] = np.fromfile(fid, dtype = np.uint16, count = 1)
        h['x_units'] = np.fromfile(fid, dtype = np.uint16, count = 1)
        h['y_units'] = np.fromfile(fid, dtype = np.uint16, count = 1)
        h['z_units'] = np.fromfile(fid, dtype = np.uint16, count = 1)
        h['t_units'] = np.fromfile(fid, dtype = np.uint16, count = 1)
        h['spare02'] = np.fromfile(fid, dtype = np.int16, count = 1)
        h['x_return_speed'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['y_return_speed'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['z_return_speed'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['scan_orientation'] = np.fromfile(fid, dtype = np.int16, count = 1)
        h['scan_direction'] = np.fromfile(fid, dtype = np.int16, count = 1)
        h['data_storage_order'] = np.fromfile(fid, dtype = np.int16, count = 1)
        h['scanner_type'] = np.fromfile(fid, dtype = np.int16, count = 1)
        h['x_inc'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['y_inc'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['z_inc'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['t_inc'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['num_x_pts'] = np.fromfile(fid, dtype = np.int32, count = 1)
        h['num_y_pts'] = np.fromfile(fid, dtype = np.int32, count = 1)
        h['num_z_pts'] = np.fromfile(fid, dtype = np.int32, count = 1)
        h['num_t_pts'] = np.fromfile(fid, dtype = np.int32, count = 1)
        h['x_speed'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['y_speed'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['z_speed'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['x_acc'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['y_acc'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['z_acc'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['x_motor_res'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['y_motor_res'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['z_motor_res'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['x_encoder_res'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['y_encoder_res'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['z_encoder_res'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['date_processed'] = b''.join(np.fromfile(fid, dtype = 'S1', count = 8))
        h['time_processed'] = b''.join(np.fromfile(fid, dtype = 'S1', count = 8))
        h['depth_recon'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['x_max_travel'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['y_max_travel'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['elevation_offset_angle'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['roll_offset_angle'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['z_max_travel'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['azimuth_offset_angle'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['adc_type'] = np.fromfile(fid, dtype = np.int16, count = 1)
        h['spare06'] = np.fromfile(fid, dtype = np.int16, count = 1)
        h['scanner_radius'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['x_offset'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['y_offset'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['z_offset'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['t_delay'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['range_gate_start'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['range_gate_end'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['ahis_software_version'] = np.fromfile(fid, dtype = np.float32, count = 1)
        h['spare_end'] = np.fromfile(fid, dtype = np.float32, count = 10)

    return h
  
#unit test ----------------------------------
header = read_header(APS_FILE_NAME)
print(len(header))

for data_item in sorted(header):
    print ('{} -> {}'.format(data_item, header[data_item]))

83
adc_max_voltage -> [ 1.10000002]
adc_min_voltage -> [-1.10000002]
adc_type -> [17]
ahis_software_version -> [ 7.0999999]
avg_data_value -> [ 3190.42041016]
azimuth_offset_angle -> [ 0.]
band_width -> [ 30058.70898438]
comments1 -> b'                                                                                '
comments2 -> b'                                                                                '
config_type -> [2]
data_scale_factor -> [  1.80140702e-08]
data_storage_order -> [9]
data_type -> [5]
data_units -> [0]
date_modified -> b''
date_processed -> b''
depth_recon -> [ 0.]
edge_weighting -> [0]
elevation_offset_angle -> [ 0.]
energy_type -> [2]
file_type -> [12]
filename -> b'                    '
frequency -> [ 25000.]
mat_velocity -> [  3.00000000e+08]
max_data_value -> [ 65535.]
min_data_value -> [ 0.]
num_polarization_channels -> [1]
num_pts -> [1]
num_t_pts -> [16]
num_x_pts -> [512]
num_y_pts -> [660]
num_z_pts -> [1]
parent_filename -> b'                    '


In [18]:
#----------------------------------------------------------------------------------
# read_data(infile):  reads and rescales any of the four image types
#
# infile:             an .aps, .aps3d, .a3d, or ahi file
#
# returns:            the stack of images
#
# note:               word_type == 7 is an np.float32, word_type == 4 is np.uint16      
#----------------------------------------------------------------------------------

from collections import namedtuple
from warnings import warn

ScanData = namedtuple('ScanData', ['header', 'data', 'real', 'imag', 'extension'])
print(ScanData)

def read_data(infile):
    """Read any of the 4 types of image files, returns a numpy array of the image contents
    """
    _, extension = os.path.splitext(infile)
    sd_dict = {'header': None, 'data': None, 'real': None, 'imag': None, 'extension': extension}
    
    h = read_header(infile)
    sd_dict['header'] = h
    nx = int(h['num_x_pts'])
    ny = int(h['num_y_pts'])
    nt = int(h['num_t_pts'])
    with open(infile, 'rb') as fid:
        fid.seek(512) #skip header
        if extension == '.aps' or extension == '.a3daps':
            if(h['word_type']==7): #float32
                data = np.fromfile(fid, dtype = np.float32, count = nx * ny * nt)
            elif(h['word_type']==4): #uint16
                data = np.fromfile(fid, dtype = np.uint16, count = nx * ny * nt)
            data = data * h['data_scale_factor'] #scaling factor
            data = data.reshape(nx, ny, nt, order='F').copy() #make N-d image
        elif extension == '.a3d':
            if(h['word_type']==7): #float32
                data = np.fromfile(fid, dtype = np.float32, count = nx * ny * nt)
            elif(h['word_type']==4): #uint16
                data = np.fromfile(fid, dtype = np.uint16, count = nx * ny * nt)
            data = data * h['data_scale_factor'] #scaling factor
            data = data.reshape(nx, nt, ny, order='F').copy() #make N-d image
        elif extension == '.ahi':
            data = np.fromfile(fid, dtype = np.float32, count = 2* nx * ny * nt)
            data = data.reshape(2, ny, nx, nt, order='F').copy()
            real = data[0,:,:,:].copy()
            imag = data[1,:,:,:].copy()
            sd_dict['real'] = real
            sd_dict['imag'] = imag
        else:
            warn('Extension not really supported: {}'.format(extension), RuntimeWarning)
            data = None
        sd_dict['data'] = data
        
    #return ScanData(**sd_dict)
    return sd_dict

#unit test ----------------------------------
d = read_data(APS_FILE_NAME)
print(d['data'].shape)
d['data']

<class '__main__.ScanData'>
(512, 660, 16)


array([[[  4.91784112e-06,   4.89982722e-06,   5.83655856e-06, ...,
           4.82777068e-06,   5.33216462e-06,   5.29613681e-06],
        [  5.13400983e-06,   4.53954590e-06,   4.71968633e-06, ...,
           4.59358807e-06,   5.63840376e-06,   6.07074162e-06],
        [  4.16125022e-06,   4.39543328e-06,   4.68365806e-06, ...,
           4.73770069e-06,   5.18805246e-06,   4.57557371e-06],
        ..., 
        [  6.07074162e-06,   6.34095250e-06,   4.48550327e-06, ...,
           3.99912369e-06,   4.32337674e-06,   5.80053074e-06],
        [  5.09798201e-06,   5.40422116e-06,   4.62961589e-06, ...,
           4.25132066e-06,   3.90905325e-06,   5.51230551e-06],
        [  5.17003809e-06,   4.64763025e-06,   3.80096890e-06, ...,
           3.78295476e-06,   4.62961589e-06,   4.89982722e-06]],

       [[  5.07996765e-06,   4.91784112e-06,   5.17003809e-06, ...,
           6.46705121e-06,   4.28734893e-06,   5.69244639e-06],
        [  7.06151559e-06,   5.49429160e-06,   6.59314992e-0

In [48]:
from keras.utils.np_utils import to_categorical

label_df = pd.read_csv(os.path.join(base_dir, 'stage1_labels.csv'))
label_df['ImageId'] = label_df['Id'].map(lambda x: x.split('_')[0])
label_df['ImageZoneId'] = label_df['Id'].map(lambda x: int(x.split('_')[1][4:]))
# create a vector with each category being an image zone
label_df['ImageZoneVec'] = label_df.apply(
    lambda c_row:[c_row['Probability']*to_categorical(c_row['ImageZoneId']-1, 
                                                      num_classes = label_df['ImageZoneId'].max())],axis=1)

In [41]:
label_df.head()

Id  Probability  \
0   00360f79fd6e02781457eda48f85da90_Zone1            0   
1  00360f79fd6e02781457eda48f85da90_Zone10            0   
2  00360f79fd6e02781457eda48f85da90_Zone11            0   
3  00360f79fd6e02781457eda48f85da90_Zone12            0   
4  00360f79fd6e02781457eda48f85da90_Zone13            0   

                            ImageId  ImageZoneId  \
0  00360f79fd6e02781457eda48f85da90            1   
1  00360f79fd6e02781457eda48f85da90           10   
2  00360f79fd6e02781457eda48f85da90           11   
3  00360f79fd6e02781457eda48f85da90           12   
4  00360f79fd6e02781457eda48f85da90           13   

                                        ImageZoneVec  
0  [[[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0...  
1  [[[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0...  
2  [[[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0...  
3  [[[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0...  
4  [[[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0...

In [34]:
label_df.Id[19478]

'ff9c9b7de5dacc8315e2bbc18c451c49_Zone6'

In [40]:
label_df.ImageZoneVec[19478][0].sum()

1.0

In [49]:
label_df.ImageZoneVec.shape

(19499,)

## Additional processing

The `.aps` version of the scans are images (sometimes refered to herein also as slices) with views from shots taken at a regular interval of $22.5$-degree rotation, $360$ degrees around the subject scanned. So for each subject, we'll get $16$ scans. Each scan will contain a view from a given angle of any visible threat zones.

Part of my approach (inspire by Brian Farrar) in this notebook is to isolate each individual threat zone from every visible angle. Later, I want to be able to make features out of each individual threat zone from each angle that a given threat zone is visible. That will allow me to later train on each threat zone individually from every view in a 2D format.

This is done by dividing a full image into sectors - this will enable me to isolate each threat zone (there are 16 threat zones). Note that each image is $512$ X $660$ pixels. 